ARTEMIS-3243 Improve Mirror with dual mirror (#3633)
apache/activemq-artemis: docs/user-manual/en/amqp-broker-connections.md

The following types of operations are supported on an AMQP server connection:

* Mirrors
  * The broker uses an AMQP connection to another broker and duplicates messages and sends acknowledgements over the wire.
* Senders
  * Messages received on specific queues are transferred to another endpoint.
* Receivers
  * The broker pulls messages from another endpoint.
* Peers
  * The broker creates both senders and receivers on another endpoint that knows how to handle them. This is currently implemented by Apache Qpid Dispatch.
    
# Reconnecting and Failover

It is possible to determine how reconnection will happen on a broker connection. 

These are the attributes are available on amqp-connection XML element:
- reconnect-attempts: default as -1 (infinite). How many attempts will be done after a failed connection
- retry-interval: default as 5000, in milliseconds, the wait between each retry in connections. 

It is also possible to specify alternate hosts on a broker connection by appending a comma separated list after a # at the end of the URI.
The broker connection would keep trying on the alternate list until one of the targets is available to connect. Example:

> ![Broker Connection Failover](images/broker-connection-failover.jpg)

Author: it's an both ways arrow actually...

> ![Broker Connection Failover](images/broker-connection-failover.jpg)

Author: Ohh... I see.. it's the failover.. you're right

### Mirror configuration:

The following optional arguments can be utilized:

* `queue-removal`: Specifies whether a queue- or address-removal event is sent. The default value is `true`.
* `message-acknowledgements`: Specifies whether message acknowledgements are sent. The default value is `true`.
* `queue-creation`: Specifies whether a queue- or address-creation event is sent. The default value is `true`.

An example of a mirror configuration is shown below:
```xml
<broker-connections>
  <amqp-connection uri="tcp://MY_HOST:MY_PORT" name="my-mirror">
    <mirror  queue-removal="true" queue-creation="true" message-acknowledgements="true" source-mirror-address="myLocalSNFMirrorQueue"/>
  </amqp-connection>
</broker-connections>
```

### Store and Forward Queue

Mirror events are always stored on a local queue prefixed as "$ACTIVEMQ_ARTEMIS_MIRROR_" concatenated with broker connection's name.

So, in the following configuration mirror events will be stored on a queue named "$ACTIVEMQ_ARTEMIS_MIRROR_my-mirror".

```xml
<broker-connection>
  <amqp-connection uri="tcp://brokerB:5672" name="brokerB">
    <mirror/>
  </amqp-connection>

> </broker-connection>

Author: I did not commit the suggestion by I imported the changes manually..

```

> These messages are then transferred to brokerB:5672. A link with the address name $ACTIVEMQ_ARTEMIS_MIRROR_brokerB will be created towards brokerB. If there is a security manager configured, security roles must be provided to the user on the broker connection.
> 
> Notice the queue $ACTIVEMQ_ARTEMIS_MIRROR_brokerB will not physically exist on brokerB and there won't be a way to find it on the administrator console. The broker will treat these messages accordingly as mirror events and will execute the appropriate operations at the target broker.

Reviewer: ```suggestion
These messages are then transferred to brokerB:5672. A producer to the address $ACTIVEMQ_ARTEMIS_MIRROR_brokerB will be created towards brokerB. If there is a security manager configured, security roles must be provided to the user on the broker connection.

Notice the queue $ACTIVEMQ_ARTEMIS_MIRROR_brokerB will not actually exist on brokerB and so it wont be visible on the administration console. The target broker will treat these messages accordingly as mirror events and perform the appropriate operations at the target broker.
```

### Pre Existing Messages

> The brokers will only send messages from the point in time the mirror was configured. Previously existing messages will not be forwarded to other brokers.

Reviewer: ```suggestion
The broker will only mirror messages arriving from the point in time the mirror was configured. Previously existing messages will not be forwarded to other brokers.
```

## Dual Mirror (Disaster and Recovery option)

ActiveMQ Artemis supports automatic failback mirroring. Every sent message and every acknowledgement is asynchronously replicated to the mirrored broker.

On the following diagram, there will be two servers called DataCenter1, and DataCenter2. In order to have a dual mirror configuration, it is necessary is to add the mirror broker connection on each broker.xml:

> ![Broker Connection DR](images/broker-connection-DR.jpg)

Reviewer: I would personally add two lines with a single arrow heads at opposoding ends, rather than 1 line with arrow heads at both ends, its more descriptive of what occurs. Again I would remove the config from the diagrams.

> Figure 2. Broker Connection Disaster and Recovery.

Reviewer: ```suggestion
Figure 2. Broker Connection - Disaster Recovery.
```

> The broker connection will replicate sends and acknowledgements no matter where they originated. If messages are sent on DC1 (DataCenter1) these will be automatically transitioned to DC2 (DataCenter2). Messages Acknowledged on DC2 will be transitionally back to DC1. 
> The only exception to that rule would be if there were consumers with pending messages on any server. It is recommended to not have active consumers on both servers.

Reviewer: ```suggestion
The broker connections will replicate sends and acknowledgements to the other broker, no matter where they originated. If messages are sent on DC1 (DataCenter1) these will be automatically transferred to DC2 (DataCenter2). Messages acknowledgements received on DC2 will be automatically related back to DC1. 
The only exception to that rule would be if there were already consumers with pending messages on any server, where a mirrored acknowledgement will not prevent the message being consumed by both consumers. It is recommended to not have active consumers on both servers.
```

## Example

> There is an example as part of the distribution showing dual broker configuration (or disaster and recovery) under ./examples/features/broker-connection/disaster-recovery.

Reviewer: ```suggestion
There is an example as part of the distribution showing dual broker configuration (or disaster recovery) under ./examples/features/broker-connection/disaster-recovery.
```

On the example two brokers are configured to mirror each other and whatever happens in one broker is immediately copied over to the other broker.

## Peers

> The broker can be configured as a peer which connects to the [Apache Qpid Dispatch Router](https://qpid.apache.org/components/dispatch-router/) and instructs it that the broker will act as a store-and-forward queue for a given AMQP waypoint address configured on the router. In this scenario, clients connect to a router to send and receive messages using a waypointed address, and the router routes these messages to or from the queue on the broker.

Reviewer: I would suggest that this entire 'Peers' section be moved later after the "## Senders and Receivers" section, as its probably the less likely to be used, and it would better match up with the order of their outline in the doc intro.

Author: good idea.. will do

The peer configuration causes ActiveMQ Artemis to create a sender and receiver pair for each destination matched in the broker-connection configuration, with these carrying special configuration to let Qpid Dispatch know to collaborate with the broker. This replaces the traditional need of a router-initiated connection and auto-links.

Qpid Dispatch Router offers a lot of advanced networking options that be used together with ActiveMQ Artemis.

With a peer configuration, the same properties are present as when there are senders and receivers. For example, a configuration where queues with names beginning "queue." act as storage for the matching router waypoint address would be:

```xml
<broker-connections>

>   <amqp-connection uri="tcp://MY_HOST:MY_PORT" name="my-router">

Reviewer: ```suggestion
  <amqp-connection uri="tcp://HOST:PORT" name="router">
```

*Notice:* If you disable auto-start on the broker connection, the start of the broker connection will only happen after the management method `startBrokerConnection(connectionName)` is called on the ServerController.

> *Important*: In addition to a matching mirror ID, the target endpoint needs permission for all operations that you configure. Therefore, If you are using a security manager, ensure that you perform the configured operations as a user with sufficient permissions.

Author: this will be updated.. I removed some code that needed matching between source and target.

I'm still updating docs.. docs are not updated yet. (hence that's why it's a draft)

## Mirrors
The mirror option on the broker connection can capture events from the broker and pass them over the wire to another broker. This enables you to capture multiple asynchronous replicas. The following types of events are captured:

* Message routing
* Message acknowledgement
* Queue and address creation
* queue and address deletion

When you configure a mirror, these events are captured from the broker, stored on a local queue, and later forwarded to a target destination on another ActiveMQ Artemis broker.

To configure a mirror, you add a `<mirror>` element within the `<amqp-connection>` element.

The local queue is called `source-mirror-address`

You can specify the following optional arguments.

* `queue-removal`: Specifies whether a queue- or address-removal event is sent. The default value is `true`.
* `message-acknowledgements`: Specifies whether message acknowledgements are sent. The default value is `true`.
* `queue-creation`: Specifies whether a queue- or address-creation event is sent. The default value is `true`.

> * `source-mirror-address`: By default, the mirror creates a non-durable temporary queue to store messages before they are sent to the other broker. If you define a name value for this property, an ANYCAST durable queue and address is created with the specified name.

Author: docs are out of sync still. I will add a commit with doc changes.

An example of a mirror configuration is shown below:
```xml
<broker-connections source-mirror-id="1">
    <amqp-connection uri="tcp://MY_HOST:MY_PORT" name="my-mirror">
            <mirror  queue-removal="true" queue-creation="true" message-acknowledgements="true" source-mirror-address="myLocalSNFMirrorQueue"/>

>             <mirror target-mirror-id="2"/>

Author: this will be removed.. docs are out of sync.

    </amqp-connection>
</broker-connections>
```

> *Important*: A broker can mirror to multiple replicas (1 to many). However a replica broker can only have a single mirror source. Make sure you do not mirror multiple source brokers to a single replica broker.

Author: @gemmellr ok, so what do you suggest? I need a broker ID as part of the ID when you have dual mirrors.

store them independently on the queue.. and make the acks using both values?

> *Important*: A broker can mirror to multiple replicas (1 to many). However a replica broker can only have a single mirror source. Make sure you do not mirror multiple source brokers to a single replica broker.

Author: this was removed from the final doc.. marking It resolved.

### Pre Existing Messages
The broker will not send pre existing messages through the mirror. So, If you add mirror to your configuration and the journal had pre existing messages these messages will not be sent. 

## Broker Connection Stop and Disconnect

> Once you start the broker connection with a mirror the mirror events will always be sent to the intermediate queue configured at the `source-mirror-address`.

Author: I will ignore anything on the docs since I'm still updating docs... ;)

> Once you start the broker connection with a mirror the mirror events will always be sent to the intermediate queue configured at the `source-mirror-address`.

Reviewer: Fair enough...though note the comment 2 down about the dodgy acceptor autostart config.

*Important:* Do not use this feature to connect to another broker, otherwise any message sent will be immediately ready to consume creating an infinite echo of sends and receives.

*Important:* It is not necessary to configure the router with a connector or auto-links to communicate with the broker. The brokers peer configuration replaces these aspects of the router waypoint usage.

## Address Consideration
It is highly recommended that `address name` and `queue name` are the same. When a queue with its distinct name (as in the following example) is used, senders and receivers will always use the `address name` when creating the remote endpoint.

```xml
<broker-connections>

>   <amqp-connection uri="tcp://MY_HOST:MY_PORT" name="my-broker">

Reviewer: ```suggestion
  <amqp-connection uri="tcp://HOST:PORT" name="other-server">
```
